File: backend/app/services/llm_service.py
"""
LLM Service
Handles communication with the LLM provider (Groq Cloud API or local Ollama).
Supports automatic fallback models to guarantee 100% uptime with zero lag.
Features:
 - Per-model connection timeout (15s connect, 30s read)
 - Proper httpx.ConnectError / httpx.TimeoutException handling
 - Yields nothing (raises RuntimeError) if ALL models fail so RAG fallback triggers
"""

import json
import logging
import httpx
from typing import AsyncGenerator
from app.config import settings

logger = logging.getLogger(__name__)

# Shared timeout configuration
_CONNECT_TIMEOUT = 15.0   # Max wait to establish TCP connection
_READ_TIMEOUT = 30.0      # Max wait between streamed chunks
_POOL_TIMEOUT = 10.0      # Max wait for a connection from the pool


class LLMService:
    """
    Unified LLM service that supports both Groq (cloud) and Ollama (local).
    Features automatic fallback model chain across all active Groq models.
    """

    def __init__(self):
        self.provider = settings.llm_provider  # "groq" or "ollama"

        if self.provider == "groq":
            self.api_key = settings.groq_api_key
            self.base_url = "https://api.groq.com/openai/v1"
            self.primary_model = settings.groq_model
            # Always try llama-3.1-8b-instant first, then llama3-8b-8192, then llama-3.3-70b-versatile
            self.fallback_models = [
                "llama-3.1-8b-instant",
                "llama3-8b-8192",
                "llama-3.3-70b-versatile",
            ]
            logger.info(
                f"LLMService initialized with Groq Cloud. Primary: {self.primary_model}, Fallbacks: {self.fallback_models}"
            )
        else:
            self.base_url = settings.ollama_base_url
            self.model = settings.ollama_model
            logger.info(f"LLMService initialized with Ollama Local. Model: {self.model}")

    async def generate_response_stream(
        self,
        prompt: str,
        system_message: str,
        history: list[dict] | None = None,
    ) -> AsyncGenerator[str, None]:
        """
        Sends a prompt and system message to the configured LLM provider
        and yields the response incrementally.
        """
        if self.provider == "groq":
            # Build list of models to try in sequence without duplicates
            models_to_try = []
            for m in [self.primary_model] + self.fallback_models:
                if m not in models_to_try:
                    models_to_try.append(m)

            success = False
            for model_name in models_to_try:
                logger.info(f"Attempting response using Groq model: {model_name}")
                received_any_chunk = False

                try:
                    async for chunk in self._stream_groq_single_model(
                        prompt, system_message, history, model_name=model_name
                    ):
                        received_any_chunk = True
                        yield chunk

                    if received_any_chunk:
                        success = True
                        break
                except (httpx.ConnectError, httpx.ConnectTimeout) as e:
                    logger.warning(f"Groq connection failed for '{model_name}': {e}. Trying fallback...")
                    continue
                except httpx.TimeoutException as e:
                    logger.warning(f"Groq timeout for '{model_name}': {e}. Trying fallback...")
                    continue
                except httpx.HTTPStatusError as e:
                    status = e.response.status_code if e.response else 0
                    logger.warning(f"Groq model '{model_name}' HTTP {status}: {e}. Trying fallback...")
                    # Don't retry on 401 (bad API key) — it will fail for all models
                    if status == 401:
                        break
                    continue
                except Exception as e:
                    logger.warning(f"Groq Model '{model_name}' unexpected error: {e}. Trying fallback...")
                    continue

            if not success:
                logger.error("All Groq models failed. Triggering Direct Knowledge Fallback...")
                raise RuntimeError("ALL_LLM_MODELS_FAILED")
        else:
            async for chunk in self._stream_ollama(prompt, system_message, history):
                yield chunk

    # ──────────────────────────────────────────────
    #  Groq Cloud API Single Model Streamer
    # ──────────────────────────────────────────────
    async def _stream_groq_single_model(
        self,
        prompt: str,
        system_message: str,
        history: list[dict] | None = None,
        model_name: str = "llama-3.1-8b-instant",
    ) -> AsyncGenerator[str, None]:
        """Stream response from a specific Groq model."""

        messages = [{"role": "system", "content": system_message}]

        if history:
            messages.extend(history)

        messages.append({"role": "user", "content": prompt})

        payload = {
            "model": model_name,
            "messages": messages,
            "stream": True,
            "temperature": 0.3,
            "max_tokens": 1200,
        }

        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

        url = f"{self.base_url}/chat/completions"

        timeout = httpx.Timeout(
            connect=_CONNECT_TIMEOUT,
            read=_READ_TIMEOUT,
            write=10.0,
            pool=_POOL_TIMEOUT,
        )

        async with httpx.AsyncClient(timeout=timeout) as client:
            async with client.stream("POST", url, json=payload, headers=headers) as response:
                if response.status_code != 200:
                    await response.aread()
                    logger.error(
                        f"Groq API model '{model_name}' status error: {response.status_code} - {response.text}"
                    )
                    raise httpx.HTTPStatusError(
                        f"Status {response.status_code}",
                        request=response.request,
                        response=response,
                    )

                async for line in response.aiter_lines():
                    if not line or not line.startswith("data: "):
                        continue
                    data_str = line[6:]
                    if data_str.strip() == "[DONE]":
                        break
                    try:
                        data = json.loads(data_str)
                        delta = data.get("choices", [{}])[0].get("delta", {})
                        content = delta.get("content", "")
                        if content:
                            yield content
                    except json.JSONDecodeError:
                        continue

    # ──────────────────────────────────────────────
    #  Ollama Local API
    # ──────────────────────────────────────────────
    async def _stream_ollama(
        self,
        prompt: str,
        system_message: str,
        history: list[dict] | None = None,
    ) -> AsyncGenerator[str, None]:
        """Stream response from local Ollama instance."""

        messages = [{"role": "system", "content": system_message}]

        if history:
            messages.extend(history)

        messages.append({"role": "user", "content": prompt})

        payload = {
            "model": self.model,
            "messages": messages,
            "stream": True,
        }

        url = f"{self.base_url}/api/chat"

        try:
            async with httpx.AsyncClient(timeout=120.0) as client:
                async with client.stream("POST", url, json=payload) as response:
                    response.raise_for_status()
                    async for chunk in response.aiter_lines():
                        if not chunk:
                            continue
                        try:
                            data = json.loads(chunk)
                            if "message" in data and "content" in data["message"]:
                                yield data["message"]["content"]
                        except json.JSONDecodeError:
                            continue
        except Exception as e:
            logger.exception("Error during Ollama generation")
            # Raise to trigger Direct Knowledge Fallback in rag_service.py
            raise RuntimeError(f"OLLAMA_FAILED: {e}")

File: php-backend/includes/LlmService.php
<?php
/**
 * LlmService — Groq Cloud API Client
 * 
 * Handles SSE streaming from the Groq API using cURL.
 * Features:
 *  - Automatic model fallback chain (3 models)
 *  - Granular timeouts (connect / read)
 *  - Proper SSE frame parsing from Groq response
 *  - 401 (bad API key) short-circuit — stops retrying immediately
 */

class LlmService
{
    private string $apiKey;
    private string $apiUrl;
    private string $primaryModel;
    private array  $fallbackModels;
    private int    $connectTimeout;
    private int    $readTimeout;
    private float  $temperature;
    private int    $maxTokens;

    public function __construct(array $config)
    {
        $this->apiKey         = $config['groq_api_key'];
        $this->apiUrl         = $config['groq_api_url'];
        $this->primaryModel   = $config['groq_model'];
        $this->fallbackModels = $config['groq_fallback_models'];
        $this->connectTimeout = $config['connect_timeout'];
        $this->readTimeout    = $config['read_timeout'];
        $this->temperature    = $config['temperature'];
        $this->maxTokens      = $config['max_tokens'];
    }

    /**
     * Stream a response from the Groq API with automatic model fallback.
     * Calls $onChunk($text) for each content fragment received.
     * 
     * @param string   $prompt        User's message
     * @param string   $systemMessage System prompt with context
     * @param array    $history       Conversation history
     * @param callable $onChunk       Callback: function(string $text): void
     * @return bool    True if any chunk was delivered, false if all models failed
     * @throws RuntimeException If ALL models fail (triggers Direct Knowledge Fallback)
     */
    public function streamResponse(
        string   $prompt,
        string   $systemMessage,
        array    $history,
        callable $onChunk
    ): bool {
        // Build deduplicated model list: primary first, then fallbacks
        $modelsToTry = [];
        foreach (array_merge([$this->primaryModel], $this->fallbackModels) as $m) {
            if (!in_array($m, $modelsToTry)) {
                $modelsToTry[] = $m;
            }
        }

        foreach ($modelsToTry as $modelName) {
            error_log("[LlmService] Attempting Groq model: {$modelName}");

            try {
                $received = $this->streamSingleModel(
                    $prompt, $systemMessage, $history, $modelName, $onChunk
                );
                if ($received) {
                    return true;
                }
            } catch (\RuntimeException $e) {
                $msg = $e->getMessage();
                error_log("[LlmService] Model '{$modelName}' failed: {$msg}");

                // Don't retry on 401 — API key is invalid for ALL models
                if (str_contains($msg, 'HTTP 401')) {
                    error_log("[LlmService] Invalid API key. Stopping all retries.");
                    break;
                }
                continue;
            }
        }

        error_log("[LlmService] ALL Groq models failed. Triggering Direct Knowledge Fallback...");
        throw new \RuntimeException('ALL_LLM_MODELS_FAILED');
    }

    /**
     * Stream response from a single Groq model using cURL.
     */
    private function streamSingleModel(
        string   $prompt,
        string   $systemMessage,
        array    $history,
        string   $modelName,
        callable $onChunk
    ): bool {
        // Build messages array
        $messages = [['role' => 'system', 'content' => $systemMessage]];
        foreach ($history as $msg) {
            $messages[] = $msg;
        }
        $messages[] = ['role' => 'user', 'content' => $prompt];

        $payload = json_encode([
            'model'       => $modelName,
            'messages'    => $messages,
            'stream'      => true,
            'temperature' => $this->temperature,
            'max_tokens'  => $this->maxTokens,
        ], JSON_UNESCAPED_UNICODE);

        $ch = curl_init();
        curl_setopt_array($ch, [
            CURLOPT_URL            => $this->apiUrl,
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => $payload,
            CURLOPT_HTTPHEADER     => [
                'Authorization: Bearer ' . $this->apiKey,
                'Content-Type: application/json',
                'Accept: text/event-stream',
                'User-Agent: InnoveraChatBot/1.0',
            ],
            CURLOPT_CONNECTTIMEOUT => $this->connectTimeout,
            CURLOPT_TIMEOUT        => $this->readTimeout + $this->connectTimeout,
            CURLOPT_RETURNTRANSFER => false,
            CURLOPT_HEADER         => false,
            CURLOPT_SSL_VERIFYPEER => false,
            CURLOPT_SSL_VERIFYHOST => 0,
            CURLOPT_IPRESOLVE      => CURL_IPRESOLVE_V4,
        ]);

        // ─── Streaming SSE Parser ───
        $buffer = '';
        $receivedAnyChunk = false;
        $httpCode = 0;
        $errorBody = '';

        curl_setopt($ch, CURLOPT_WRITEFUNCTION,
            function ($ch, $data) use (&$buffer, &$receivedAnyChunk, &$httpCode, &$errorBody, $onChunk) {
                // Get HTTP status code (available after headers received)
                if ($httpCode === 0) {
                    $httpCode = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
                }

                // If non-200 response, collect error body for logging
                if ($httpCode !== 0 && $httpCode !== 200) {
                    $errorBody .= $data;
                    return strlen($data);
                }

                $buffer .= $data;

                // Process complete lines from buffer
                while (($pos = strpos($buffer, "\n")) !== false) {
                    $line = substr($buffer, 0, $pos);
                    $buffer = substr($buffer, $pos + 1);
                    $line = trim($line);

                    if (empty($line)) continue;
                    if (!str_starts_with($line, 'data: ')) continue;

                    $dataStr = substr($line, 6);

                    if (trim($dataStr) === '[DONE]') {
                        return strlen($data);
                    }

                    $json = json_decode($dataStr, true);
                    if ($json && isset($json['choices'][0]['delta']['content'])) {
                        $content = $json['choices'][0]['delta']['content'];
                        if ($content !== '') {
                            $receivedAnyChunk = true;
                            $onChunk($content);
                        }
                    }
                }

                return strlen($data);
            }
        );

        $result = curl_exec($ch);
        $curlError = curl_error($ch);
        $curlErrno = curl_errno($ch);
        $finalHttpCode = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        // Handle cURL errors (connection, DNS, timeout)
        if ($result === false) {
            $errorType = match (true) {
                in_array($curlErrno, [CURLE_COULDNT_RESOLVE_HOST, CURLE_COULDNT_CONNECT])
                    => 'Connection failed',
                $curlErrno === CURLE_OPERATION_TIMEDOUT
                    => 'Request timeout',
                default => 'cURL error',
            };
            throw new \RuntimeException("{$errorType} for '{$modelName}': {$curlError}");
        }

        // Handle HTTP error responses
        if ($finalHttpCode !== 200 && $finalHttpCode !== 0) {
            throw new \RuntimeException("HTTP {$finalHttpCode} from Groq model '{$modelName}'");
        }

        return $receivedAnyChunk;
    }
}
